Fixes #21045 - React components for formatting dates

tstrachota opened the pull request:
Adds angular directives that use react components from the foreman to provide consistent date and time formatting. It uses ng-react for mounting the components from within angular views.

Depends on a PR into the Foreman: https://github.com/theforeman/foreman/pull/5184

waldenraines commented on engines/bastion_katello/app/assets/javascripts/bastion_katello/products/details/views/product-info.html:
```diff
@@ -63,13 +63,14 @@ <h4 translate>Sync Status</h4>
-        {{ product.last_sync_words }} Ago ({{ product.sync_status.finish_time | date:'medium' }} Local Time)
+      <dd>
+        <relative-date date="product.last_sync.ended_at" />
+        (<date date="product.last_sync.finish_time" /> <translate>Local Time</translate>)
```
I think this should just be `product.last_sync` because I am seeing `product.last_sync.finish_time` as `undefined`:

![screen shot 2017-10-12 at 11 10 13 am](https://user-images.githubusercontent.com/4116405/31503539-ef43b0fa-af3d-11e7-8a9e-4012a455fe14.png)

Perhaps you need the same logic as in `repositories-info.html`, I'm not sure what our API is up to here.

tstrachota commented:
@waldenraines you're right. I used wrong grep pattern and missed all date filter with double quotes. I'll repleace the dates also there once we decide on the formats: https://github.com/theforeman/foreman/pull/4909#issuecomment-336459752

waldenraines commented:
> @waldenraines you're right. I used wrong grep pattern and missed all date filter with double quotes. I'll repleace the dates also there once we decide on the formats: theforeman/foreman#4909 (comment)

It looks like we have reached a consensus on the date formats so I think you're good to proceed with this when you are able.

tstrachota commented:
Updated:
- added new date components from https://github.com/theforeman/foreman/pull/4909
- replaced dates with new directives on locations where it was missing

waldenraines commented:
@tstrachota can you fix the hound and test issues please?

tstrachota commented:
@waldenraines thanks for re-reviewing that. I've fixed the one issue Hound commented about before.

The test failures are caused by missing https://github.com/Katello/bastion/commit/af786aa46b4ac3cb676279ae33fa8bd676292fef in bastion. Could you release new version of bastion please? I would then bump the dependency and that should fix the tests.

waldenraines commented:
> The test failures are caused by missing Katello/bastion@af786aa in bastion. Could you release new version of bastion please? I would then bump the dependency and that should fix the tests.

Oh right.  I released a new version of bastion.  Since it's just a minor version it shouldn't require any changes in this PR.

waldenraines commented:
[test]

waldenraines commented:
Blocked by https://github.com/theforeman/foreman/pull/4909

waldenraines commented:
[test]

tstrachota commented:
Thanks for releasing that @waldenraines I bumped the dependency in the gemspec.

waldenraines commented:
@tstrachota looks like we are getting legitimate eslint errors now:

```
22:30:40 [4mRunning "eslint:target" (eslint) task[24m
22:30:45 
22:30:45 app/assets/javascripts/bastion_katello/dates/dates.module.js
22:30:45   14:4   error  Expected indentation of 4 characters   indent
22:30:45   14:9   error  "tfm" used outside of binding context  block-scoped-var
22:30:45   14:9   error  'tfm' is not defined                   no-undef
22:30:45   18:4   error  Expected indentation of 4 characters   indent
22:30:45   18:36  error  'tfm' is not defined                   no-undef
22:30:45   18:36  error  "tfm" used outside of binding context  block-scoped-var
22:30:45   24:4   error  Expected indentation of 4 characters   indent
22:30:45   24:36  error  'tfm' is not defined                   no-undef
22:30:45   24:36  error  "tfm" used outside of binding context  block-scoped-var
22:30:45   30:4   error  Expected indentation of 4 characters   indent
22:30:45   30:36  error  'tfm' is not defined                   no-undef
22:30:45   30:36  error  "tfm" used outside of binding context  block-scoped-var
22:30:45   35:4   error  Expected indentation of 4 characters   indent
22:30:45   35:36  error  'tfm' is not defined                   no-undef
22:30:45   35:36  error  "tfm" used outside of binding context  block-scoped-var
22:30:45 
22:30:45 ✖ 15 problems (15 errors, 0 warnings)
```

tstrachota commented:
Rebased

waldenraines commented:
[test]

jlsherrill commented:
@tstrachota do you mind rebasing?  I'm not sure what was failing before, but a re-run should tell us. Thanks!

waldenraines commented:
Note that the foreman changes were reverted in https://github.com/theforeman/foreman/pull/5094 so http://projects.theforeman.org/issues/21312 will need to be closed first before this can be merged.

tstrachota commented:
Updated. This PR now depends on: https://github.com/theforeman/foreman/pull/5184

waldenraines commented:
[test katello]

waldenraines commented:
[test default]

waldenraines commented:
[test default]

jlsherrill commented:
@waldenraines you can ignore the default test now.  it no longer exists, but won't go away from a PR unless you open a new one.  We've just been ignoring it on existing prs.  The katello test covers everything now.

waldenraines commented:
> @waldenraines you can ignore the default test now. it no longer exists, but won't go away from a PR unless you open a new one. We've just been ignoring it on existing prs. The katello test covers everything now.

Thanks.  Is there any way to rerun hound other than pushing again?

jlsherrill commented:
@waldenraines nope, but the katello test runs rubocop, so if that passes its fine to ignore hound.  (hound is just there to make your life easier by pointing out issues within the pr).

waldenraines commented:
@tstrachota there is one other date being used on the new debian repositories page that has not been replaced in this PR:

```
[walden@yesod katello ((1384db4a3f...))]$ grep -r 'date:"short*' engines/bastion_katello/
engines/bastion_katello//app/assets/javascripts/bastion_katello/content-views/details/views/content-view-deb-repositories.html:            {{ repository.last_sync.ended_at | date:"short" }}
```

waldenraines commented:
@tstrachota ping

tstrachota commented:
Updated

waldenraines commented:
[test katello]

beav commented:
@tstrachota it looks like https://github.com/theforeman/foreman/pull/5184#pullrequestreview-104792873 (comment in PR that blocks this one) is waiting on your input, can you take a look?

tstrachota commented:
@beav it's not only that. Unfortunately there are errors in ui tests that use the fallback intl. I wasn't able to reproduce the situation outside of phantomjs. That's the main blocker. I try to play with it time to time but without any significant progress.

beav commented:
quick status update for this pr: it looks like https://github.com/theforeman/foreman/pull/5184 is still actively being worked on, no action needed for this PR at this time

tstrachota commented:
Rebased and updated to match recent changes in the Foreman counterpart.

waldenraines commented:
[test katello]

tstrachota commented:
[test katello]

waldenraines commented:
[test katello]

waldenraines commented:
@tstrachota looks like a bunch of angular tests are failing.  Maybe a missing mock?

waldenraines commented:
> @tstrachota looks like a bunch of angular tests are failing. Maybe a missing mock?

Tomas reminded me that these tests will obviously fail until the foreman bits are merged, duh Walden

waldenraines commented:
[test katello]

waldenraines commented:
Will require https://github.com/Katello/bastion/pull/240 for green tests.

beav commented:
[test katello] other PR is merged

waldenraines commented:
> [test katello] other PR is merged

This will fail because I hadn't released bastion yet (bastion--).

waldenraines commented:
[test katello]

beav commented:
@waldenraines it looks like most of the tests failed, but I don't know if it's because of my incorrect run request earlier or if it's using the newer bastion

waldenraines commented:
[test]

waldenraines commented:
[test]

waldenraines commented:
[test]

waldenraines commented:
[test katello]

beav commented:
:fireworks:
